File: Java/DSA/DequeDLL.java
import java.util.Scanner;
class Node{
    public Node prev;
    public int item;
    public Node next;
}
class Deque{
    private Node front;
    private Node rear;
    public Deque(){
        front = null;
        rear = null;
    }
    public void insertAtFront(int data){
        Node n = new Node();
        n.prev = null;
        n.item = data;
        n.next = front;
        if(front!=null){
            front.prev = n;
        }
        else{
            rear = n;
        }
        front = n;
    }
    public void insertAtRear(int data){
        Node n = new Node();
        n.item = data;
        n.next = null;
        n.prev = rear;
        if(rear!=null){
            rear.next = n;
        }
        else{
            front = n;
        }
        rear = n;
    }
    public void deleteFront(){
        Node t;
        if(front!=null){
            t = front;
            if(front==rear){
                front=rear=null;
            }
            else{
                front = front.next;
                front.prev = null;
            }
            t = null;
        }
    }
    public void printDeque(){
        Node temp;
        temp = front;
        while(temp!=null){
            System.out.print(temp.item+" ");
            temp = temp.next;
        }
        System.out.println(" ");
    }
    public void deleteRear(){
        Node r;
        if(rear!=null){
            r = rear;
            if(rear==front){
                rear=front=null;
            }
            else{
                rear = rear.prev;
                rear.next = null;
            }
            r = null;
        }
    }
    public Node getFront(){
        return front;
    }
    public Node getRack(){
        return rear;
    }
    public int getFrontElement(){
        return front.item;
    }
    public int getRearElement(){
        return rear.item;
    }
    public boolean isEmpty(){
        return front==null;
    }
}
public class DequeDLL {
    public static void main(String[] args) {
        Deque d = new Deque();
        Scanner input = new Scanner(System.in);
        System.out.println("Enter the number for inserting the Elmenet in the Deque");
        int n = input.nextInt();
        System.out.println("Enter the Element Front in the Deque");
        for(int i =0;i<n;i++){
            int b = input.nextInt();
            d.insertAtFront(b);
        }
        input.close();
        d.printDeque();
        d.deleteFront();
        System.out.println("Element Enter by you inside an Deque");
        d.printDeque();

    }
}

File: CPP-Programing/DSA/DequeDLL.cpp
#include<iostream>
struct node{
    node *prev;
    int item;
    node *next;
};
class Deque{
    private:
       node *front;
       node *rear;
    public:
       Deque();
       void insertAtFront(int);
       void insertAtRear(int);
       void deleteFront();
       void deleteRear();
       ~Deque();
       node* getFront();
       node* getRear();
       int getFrontElement();
       int getRearElement();
       bool isEmpty();
       void printTheDeque();
};
Deque::Deque(){
    front = NULL;
    rear = NULL;
}
void Deque::printTheDeque(){
    node *temp;
    temp = front;
    while(temp!=NULL){
        std::cout<<temp->item<<" ";
        temp = temp->next;
    }
    std::cout<<" "<<std::endl;
}
bool Deque::isEmpty(){
    return front==NULL;
}
int Deque::getRearElement(){
    return rear->item;
}
int Deque::getFrontElement(){
    return front->item;
}
node* Deque::getRear(){
    return rear;
}
node* Deque::getFront(){
    return front;
}
Deque::~Deque(){
    while(front){
        deleteFront();
    }
}
void Deque::deleteRear(){
    node *ptr;
    if(rear){
        ptr = rear;
        if(rear==front){
            rear=front=NULL;
        }
        else{
            rear = rear->prev;
            rear->next = NULL;
        }
        delete ptr;
    }
}
void Deque::deleteFront(){
    node *temp;
    if(front){
        temp = front;
        if(front==rear){
            front=rear=NULL;
        }
        else{
            front = front->next;
            front->prev = NULL;
        }
        delete temp;
    }
}
void Deque::insertAtRear(int data){
    node *n = new node();
    n->prev = rear;
    n->item = data;
    n->next = NULL;
    if(rear){
        rear->next = n;
    }
    else
        front = n;
    rear = n;
}
void Deque::insertAtFront(int data){
    node *n = new node();
    n->prev = NULL;
    n->item = data;
    n->next = front;
    if(front){
        front->prev = n;
    }
    else
        rear = n;
    front = n;
}
// Deriver program for checking the code
int main(){
    Deque d;
    int N,b;
    std::cout<<"Enter the number for inserting Element in the Deque"<<std::endl;
    std::cin>>N;
    std::cout<<"Enter the Element at the front of Deque"<<std::endl;
    for(int i =0;i<N;i++){
        std::cin>>b;
        d.insertAtFront(b);
    }
    std::cout<<"Enter the Element at the rear of Deque "<<std::endl;
    for(int i =0;i<N;i++){
        std::cin>>b;
        d.insertAtRear(b);
    }
    d.printTheDeque();
    return 0;
}